### PR TITLE
3.7

### PR DESCRIPTION
**Pull request by DinoV**
This updates the compiler to support Python 3.6 and 3.7 (fixing issue #7 ).  It also implements new functionality like the peephole optimizer as well as the AST based optimizer.  

I've left the existing test cases at the root of the repo pretty much as-is, although they might just want to be deleted.  There's a new set of test cases added into test_compiler which include the existing test_corpus, some new unit tests, and some new SBS code tests which diff just portions of the opcodes so they can be more version resilient.  There's also a full pass on the entire CPython standard library which makes sure that the code produced from the compiler is 100% identical to the standard compiler.

There's no attempt to maintain compatibility with 3.5, after this only 3.6 and 3.7 are supported.

Test cases can now be run using:
```
python3 -m test_compiler 
python36 -m test_compiler
```

### CODE REVIEW

**Comment by pfalcon** [comment]
This is absolutely great, and I'm left speechless at how awesome it is!

The best and shortest thing I probably could say is: would you like to maintain this upgraded version of python-compiler going forward?

**Comment by DinoV** [comment]
Do you mean just giving us commit access?  We are actively using it so we'll certainly have more fixes going forward!

**Comment by pfalcon** [comment]
@DinoV: So, I'm trying to build a mental model of what happens here. Given that the last commit date in this patchset is Feb 2020, I may imagine it *could* be following: you got some tangential task as part of your dayjob, and that's how you hacked on python-compiler for some time. It didn't work out too much, and have been sitting in the corporate server since then. Making up the year's totals, you remembered about it and got a permission to open-source it.

So if it's **only** that, it's absolutely a great Santa's present! And then it makes sense to treat it as such, avoid long nitpicking reviews (who in their right mind nitpicks at Santa's presents), and just fall on my original reasoning behind this project "if nobody else does it, I will".

But if by any chance it's **more** than that, and you can continue working on it as part of your dayjob, then I would encourage and appreciate your taking the maintainership of it. Because it was just that: I have a strong belief that there should be Python-in-Python bytecode compiler, I had excuse to forward-port Python2's implementation, and well, I did my part of the job, stopping at Python3.5. Since then, it's available to anyone for use in its current shape, or for pick-up to move it forward. And I did all work on it as a private person in my own free time, so engineers from a well-known company with strong open-source component and culture will certainly be able to maintain in better than myself.

> Do you mean just giving us commit access? 

So if the above is the case, you're welcome to do it in any way you see fit. E.g.:

* Choice 1:
You just push it as https://github.com/facebook/python-compiler and I re-fork it, to keep whatever existing links were posted around non-broken. (I'd say name "python-compiler" is good and descriptive enough. Of course, "python-bytecode-compiler" would be even more descriptive, but also too long, why I didn't use it in the first place.)

* Choice 2:
We create a github organization, where you're admin, and I transfer this repo to it.

From an outsider's perspective, Choice 2 is better: first of all, there will be an automatic redirect, instead of just potentially stale fork. Then, https://github.com/facebook/ hosts so much cool stuff, that individual projects are easily lost in so bright sidelight. But I can understand that hosting under company's organization can be a part of brand awareness, so I would understand if the preference is to host it there. (But then I don't know if https://github.com/facebook/ is only for the Facebook-originated (and thus fully Facebook-copyrighted) projects, and outside project actually shouldn't be hosted there.)


Bottom line:

1. Please let me know what you think/decide.
2. Then let me know how I can help with the transition, assuming it happens.


Thanks!

**Comment by DinoV** [comment]
I'm going through the internal process for this now, but I think we're likely to end up at something like choice 1.  We had originally just internally requested to upstream things, so creating a new repo just requires some extra internal steps for us (which not surprisingly probably won't be done until the new year).

So to help you with your mental model, the original work was done kind of optimistically under the belief when we'd need the compiler and I had some spare time to work on it.  It's taken some time but the compiler is now something we're using internally, and it doesn't just stop on updates to Python 3.7.  That was just a good first logical chunk to upstream.  We also have updates to bring it up to add 3.8 support - though we still need to do a little bit of work to make sure our 3.7 support hasn't regressed with those changes.  

The compiler is actually now being used as the base for a couple of different internal efforts.  So we will be pretty dedicated to maintaining it but I also suspect we'll continue to lag behind major Python versions by a little bit.  So I'll work through the internal process and keep you updated as things move along.

**Comment by pfalcon** [comment]
Happy New Year!

@DinoV:

> I'm going through the internal process for this now, but I think we're likely to end up at something like choice 1.

Sounds good, please keep me posted about your progress!

> So to help you with your mental model ...

Thanks!

Another question I have is who's matthiasb, as mentioned in some Phabricator metadata, e.g. `Reviewers: matthiasb, #pyro`?

**Comment by tekknolagi** [comment]
Happy New Year! I'm Max and I'm also working on this project, though I did not author any of the commits in this PR. @matzeb is my manager. We tried to scrub the relevant metadata but clearly we missed a small chunk. Thanks for spotting it -- we'll remove it shortly.

**Comment by pfalcon** [comment]
@tekknolagi:

> I'm Max and I'm also working on this project

Hey, cool! Who would have thought! So, is LISP a retreat for you from boring dayjob work, or you might hack on Python more overall? ;-)

> @MatzeB is my manager.

Great, just like I thought. I'm a big fan of his SSA works! But yeah, I was really surprised to see a familiar username in this patchset, based on his profile page info, I thought he might be at Apple on LLVM or something. I guess, I need to straighten up my knowledge of the Bay Area geography ;-). I actually dropped him a line last summer with some stupid questions regarding SSA and Python, so I'll treat all what's happening as a kind of reply, and I really looking forward to what may surface later ;-).

**Comment by tekknolagi** [comment]
> is LISP a retreat for you from boring dayjob work

The LISP stuff is just me wanting to publicly write about runtimes and compilers and optimization. I think it's easier (policy wise) to write about something I am not directly working on for my day job. I certainly plan on hacking on Python more!

Yep, he's at FB now :)

**Comment by pfalcon** [comment]
@DinoV, @tekknolagi: Any updates?

**Comment by DinoV** [comment]
@pfalcon We've got basic approval, we're just working through getting the new repo setup and making sure we're following the various policies and attempting to bypass the ones we don't like :)

**Comment by tekknolagi** [comment]
We've navigated the system and now we have https://github.com/facebookincubator/python-compiler

**Comment by pfalcon** [comment]
Great news! I've renamed this repo as "python-compiler.old", and reforked https://github.com/pfalcon/python-compiler to point to the facebookincubator repo, and sent announcement to the python-announce-list about maintainership change.

**Comment by tekknolagi** [comment]
We can probably close this PR.
